cloudant-client/src/main/java/com/cloudant/client/api/Database.java

```diff
@@ -279,20 +281,61 @@ public Shard getShard(String docId) {
      *
      * @param indexName     optional name of the index (if not provided one will be generated)
      * @param designDocName optional name of the design doc in which the index will be created
-     * @param indexType     optional, type of index (only "json" as of now)
+     * @param indexType     optional, type of index (only "json" for this method)
      * @param fields        array of fields in the index
+     * @see Database#createIndex(String)
      */
+    @Deprecated
     public void createIndex(String indexName, String designDocName, String indexType,
                             IndexField[] fields) {
-        JsonObject indexDefn = getIndexDefinition(indexName, designDocName, indexType, fields);
-        createIndex(indexDefn.toString());
+        if (indexType == null || "json".equalsIgnoreCase(indexType)) {
+            JsonIndex.Field[] jFields = new JsonIndex.Field[fields.length];
+            int i = 0;
+            for (IndexField f : fields) {
+                Order order;
+                switch (f.getOrder()) {
+                    case desc:
+                        order = Order.DESC;
+                        break;
+                    case asc:
+                    default:
+                        order = Order.ASC;
+                        break;
+                }
+                jFields[i] = new JsonIndex.Field(f.getName(), order);
+                i++;
+            }
+            createIndex(new JsonIndex.Builder().fields(jFields).definition());
+        } else {
+            throw new CouchDbException("Unsupported index type " + indexType);
+        }
     }
 
     /**
-     * Create a new index from a JSON string
+     * <P>
+     * Create a new index from a string of JSON representing the index definition
+     * </P>
+     * <P>
+     * Helpers are available to construct the index definition string for JSON and text indexes.
+     * </P>
+     * <P>
+     * Example usage:
+     * </P>
+     * <pre>
+     * {@code
+     * // Create a JSON index with generated names for the field named "Movie_year" with the default
+     * // ascending sort order
+     * db.createIndex(new JsonIndex.Builder().fields(new JsonIndex.Field("Movie_year")).definition());
+     *
+     * // Create a text index for the string field named "Movie_title"
```

Comment:
I initially had statics on the `Builder` superclass as short cuts for instantiating the builder types, but it didn't make sense when it moved into the internal package. If we had Java 8 minimum we could add static methods to the interface to instantiate the builders e.g.:
`Index.json().fields(...)`
`Index.text().feilds(...)`
otherwise we'll have to add another class or add it to each index type
`JsonIndex.configure/define/create/builder().fields(...)`

I tried to stay close to the JSON form (as we did with the proposed QueryBuilder). I can definitely see the plus sides of being less like the JSON and easier to use so if we're going to move away from that I'd be inclined to go even further and just move the `asc` `desc` etc onto the `Builder` and make the `fields(...)` method internal e.g.

`JsonIndex.builder().asc("Movie_year").done()`
`TextIndex.builder().string("Movie_title").done()`

Reply:
OK so maybe the internal/external API split isn't quite right. I appreciate that you don't want to create even more classes but it doesn't seem quite right to me that `Builder.newInstance()` actually returns (eg) a `JsonIndex`, which is doing double-duty as builder pattern class model/value object.

The naming a bit misleading here, your `Builder` is really a factory of sorts, and the thing it instantiates is a builder of sorts. I just wonder if this isn't very intuitive to an API user.

Reply:
The `newInstance()` is entirely internal and not visible to the API user, it's a utility for creating an instance of the correct type of the thing that needs to be built.

Reply:
Sorry, I got confused about `newInstance`, now I understand what that does.

Here is a suggestion for a different way of constructing `Builder`s without the user having to directly know about the internal static `Builder` class. It's only a suggestion and you might not think it's worth the extra code.

Reply:
Forgot to add the actual link - [here](https://github.com/cloudant/java-cloudant/pull/382/files)

Reply:
Done in 1abf58a
`JsonIndex.builder()`
`TextIndex.builder()`
to avoid the new'ing up

Reply:
Done in b1a418a
Changed to builder field name methods `asc` `desc` `string` `number` `bool`
each being a var args of field names, but can be called multiple times as well.